From 1ef1d2194bbf1fe8ea82390bbd786dfee380740f Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Wed, 23 Jul 2025 21:56:28 -0400
Subject: [PATCH 221/365] ntoskrnl: Implement MmGetPhysicalMemoryRanges.

probably can be moved to dll attach
---
 dlls/ntoskrnl.exe/ntoskrnl.c | 16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/dlls/ntoskrnl.exe/ntoskrnl.c b/dlls/ntoskrnl.exe/ntoskrnl.c
index 7536f65a405..733a212bfba 100644
--- a/dlls/ntoskrnl.exe/ntoskrnl.c
+++ b/dlls/ntoskrnl.exe/ntoskrnl.c
@@ -3129,12 +3129,18 @@ PHYSICAL_ADDRESS WINAPI MmGetPhysicalAddress(void *virtual_address)
 
 PHYSICAL_MEMORY_RANGE *WINAPI MmGetPhysicalMemoryRanges(void)
 {
-    static PHYSICAL_MEMORY_RANGE range = {
-        .BaseAddress.QuadPart = 0xdead,
-        .NumberOfBytes.QuadPart = 0x10000000
-    };
+    static volatile LONG once;
+    static PHYSICAL_MEMORY_RANGE range;
+    SYSTEM_BASIC_INFORMATION info;
 
-    FIXME("stub!\n");
+    TRACE("\n");
+
+    if (!InterlockedCompareExchange(&once, 1, 0))
+    {
+        NtQuerySystemInformation(SystemBasicInformation, &info, sizeof(info), NULL);
+        range.BaseAddress.QuadPart = info.MmLowestPhysicalPage;
+        range.NumberOfBytes.QuadPart = info.MmNumberOfPhysicalPages * info.PageSize;
+    }
 
     return &range;
 }
-- 
2.50.1

